[BUG] Fail to compare <<UNORDERED>> arrays where elements have <<PRESENCE>>

### What exactly did you do?

Compare `<<UNORDERED>>` JSON arrays where the elements are objects and the expectation for some fields only checks for `<<PRESENCE>>`.

Test example:

```go
func TestUNORDEREDwithPRESENCE(t *testing.T) {
	act := `{
		"data": [
			{ "foo": 1, "bar": 2 },
			{ "foo": 11, "bar": 22 }
		]
	}`
	exp := `{
		"data": [
			"<<UNORDERED>>",
			{ "foo": 11, "bar": "<<PRESENCE>>" },
			{ "foo": 1, "bar": "<<PRESENCE>>" }
		]
	}`
	ja := jsonassert.New(t)
	ja.Assertf(act, exp)
}
```

See test examples with some sanity in [this Go Playground](https://go.dev/play/p/OTV-Bf3ZOLk).

### What did you expect would happen?

The test should pass - I should be able to compare `<<UNORDERED>>` arrays having elements with `<<PRESENCE>>` in them.

### What actually happened?

```console
=== RUN   TestUNORDEREDwithPRESENCE
    prog.go:81: actual JSON at '$.data[0]' contained an unexpected element: {"bar":2,"foo":1}
    prog.go:81: actual JSON at '$.data[1]' contained an unexpected element: {"bar":22,"foo":11}
    prog.go:81: expected JSON at '$.data[0]':
        {"bar":"\u003c\u003cPRESENCE\u003e\u003e","foo":11}
        was missing from actual payload
    prog.go:81: expected JSON at '$.data[1]':
        {"bar":"\u003c\u003cPRESENCE\u003e\u003e","foo":1}
        was missing from actual payload
--- FAIL: TestUNORDEREDwithPRESENCE (0.00s)
FAIL
```

### Additional info

- Output from `go version`: `go version go1.18.2 darwin/arm64`
- Version of this package: `v.1.1.0`

As can be seen in the [Go Playground example](https://go.dev/play/p/OTV-Bf3ZOLk), I did some sanity tests, and it seems that   `<<UNORDERED>>` and `<<PRESENCE>>` work well for the example above, **except** for when they are mixed together.
Looking in the code, it seems to be caused by. the difference between [`Asserter.checkArrayUnordered`](https://github.com/kinbiko/jsonassert/blob/main/array.go#L52) and [`Asserter.checkArrayOrdered`](https://github.com/kinbiko/jsonassert/blob/main/array.go#L88). The unordered version uses [`Asserter.deepEqual`](https://github.com/kinbiko/jsonassert/blob/main/array.go#L65), whereas the ordered version uses [`Asserter.pathassertf`](https://github.com/kinbiko/jsonassert/blob/main/core.go#L13).

--- a/array.go
+++ b/array.go
@@ -49,7 +49,7 @@ func (a *Asserter) checkArrayUnordered(path string, act, exp []interface{}) {
 	for i, expEl := range exp {
 		found := false
 		for _, actEl := range act {
-			found = found || a.deepEqual(expEl, actEl)
+			found = found || a.deepEqual(actEl, expEl)
 		}
 		if !found {
 			serializedEl := serialize(expEl)
@@ -60,16 +60,6 @@ func (a *Asserter) checkArrayUnordered(path string, act, exp []interface{}) {
 			}
 		}
 	}
-}
-
-func (a *Asserter) deepEqual(act, exp interface{}) bool {
-	// There's a non-zero chance that JSON serialization will *not* be
-	// deterministic in the future like it is in v1.16.
-	// However, until this is the case, I can't seem to find a test case that
-	// makes this evaluation return a false positive.
-	// The benefit is a lot of simplicity and considerable performance benefits
-	// for large nested structures.
-	return serialize(act) == serialize(exp)
 }
 
 func (a *Asserter) checkArrayOrdered(path string, act, exp []interface{}) {

--- a/helper.go
+++ b/helper.go
@@ -12,3 +12,18 @@ type noopHelperTT struct {
 func (*noopHelperTT) Helper() {
 	// Intentional NOOP
 }
+
+// deepEqualityPrinter is a utility Printer that lets the jsonassert package
+// verify equality internally without affecting the external facing output,
+// e.g. to verify where one potential candidate is matching or not.
+type deepEqualityPrinter struct{ count int }
+
+func (p *deepEqualityPrinter) Errorf(msg string, args ...interface{}) { p.count++ }
+func (p *deepEqualityPrinter) Helper()                                { /* Intentional NOOP */ }
+
+func (a *Asserter) deepEqual(act, exp interface{}) bool {
+	p := &deepEqualityPrinter{}
+	deepEqualityAsserter := &Asserter{tt: p}
+	deepEqualityAsserter.pathassertf("", serialize(act), serialize(exp))
+	return p.count == 0
+}

--- a/helper.go
+++ b/helper.go
@@ -22,7 +22,7 @@ func (p *deepEqualityPrinter) Errorf(msg string, args ...interface{}) { p.count+
 func (p *deepEqualityPrinter) Helper()                                { /* Intentional NOOP */ }
 
 func (a *Asserter) deepEqual(act, exp interface{}) bool {
-	p := &deepEqualityPrinter{}
+	p := &deepEqualityPrinter{count: 0}
 	deepEqualityAsserter := &Asserter{tt: p}
 	deepEqualityAsserter.pathassertf("", serialize(act), serialize(exp))
 	return p.count == 0

--- a/integration_test.go
+++ b/integration_test.go
@@ -306,6 +306,12 @@ but expected JSON was:
 					`["<<UNORDERED>>", {"1": 1}, {"2": 2}, {"3": 3}, {"4": 4}, {"5": 5}, {"6": 6}, {"7": 7}, {"8": 8}, {"9": 9}, {"10": 10}, {"11": 11}, {"12": 12}, {"13": 13}, {"14": 14}, {"15": 15}, {"16": 16}, {"17": 17}, {"18": 18}, {"19": 19}, {"20": 20}]`,
 					nil,
 				},
+				"unordered array with objects with PRESENCE directives": {
+					// Regression test for https://github.com/kinbiko/jsonassert/issues/39
+					`{ "data": [ { "foo": 1, "bar": 2 }, { "foo": 11, "bar": 22 } ] }`,
+					`{ "data": [ "<<UNORDERED>>", { "foo": 11, "bar": "<<PRESENCE>>" }, { "foo": 1, "bar": "<<PRESENCE>>" } ] }`,
+					nil,
+				},
 			} {
 				tc := tc
 				t.Run(name, func(t *testing.T) { tc.check(t) })

